vllm/worker/model_runner.py

```diff
@@ -405,6 +406,10 @@ def _prepare_sample(
                 selected_token_indices.append(selected_token_start_idx +
                                               subquery_len - 1)
                 selected_token_start_idx += max_subquery_len
+
+                if sampling_params.seed is not None:
+                    sampling_params._generator = torch.Generator(
```

Comment:
If I read the code correctly, we'll be initializing a new generator every iteration with the same seed, meaning the sampling for each token will be the same. This can lead to issues such as a token being sampled repeatedly. I think it would be a good idea to deterministically permute the seed based on the number of tokens in sequence (prompt+generated).

Reply:
> we'll be initializing a new generator every iteration with the same seed,

@Yard1 I don't think so since this is within the `if seq_group_metadata.is_prompt:` block. The generator is stashed in the `sampling_params` object and will be re-used on subsequent iterations (see the change a few lines down in the corresponding `else` block).

I stashed in `sampling_params` since this was the only accessible "persistent" object tied to the sequence group, but just realized that this is a mistake since I assume that's the same object passed in the API and it could be reused across requests. I'll make a change to address this now!

Reply:
hmm but doesn't the sampling params get regenerated every iteration? furthermore, we want to be able to support preempted sequences as well

Reply:
No the SamplingParams are created once (you may be thinking of SequenceGroupMetadata), and is referenced from the SequenceGroup object, which persists even across such preemptions.

In any case for the other reason mentioned this is a bad place to store the generator and I've pushed another [commit](https://github.com/vllm-project/vllm/pull/2514/commits/be454204b8d0b2a100d4c5bc003d3fa01ba0fcc6) to add a `state` dict to the SequenceGroup to use for this instead.

I'll add the deterministic sampling test next.

Reply:
Yeah my bad, I got it confused.

We should make sure the SamplingParams are deepcopied in `add_request` to ensure that the state is not being overwritten across requests.

Reply:
@Yard1 I opened a separate PR for this since it's no longer applicable to this PR https://github.com/vllm-project/vllm/pull/2881